File: JavaScript/Tree/Heap.js
class Heap {
    constructor(type, array) {
        this.type = type

        if (array) {
            this.array = array
            this.build()
        } else {
            this.array = []
        }
    }

    swap(parent, child) {
        [this.array[parent], this.array[child]] = [this.array[child], this.array[parent]]
    }

    inBounds(index) {
        return index >= 0 && index < this.array.length
    }

    getLeftChild(index) {
        let left = index * 2 + 1
        return this.inBounds(left) ? left : -1
    }

    getRightChild(index) {
        let right = index * 2 + 2
        return this.inBounds(right) ? right : -1
    }

    getParent(index) {
        let parent = Math.floor((index - 1) / 2)
        return this.inBounds(parent) ? parent : -1
    }

    getBiggestChild(index) {
        let biggest = index
        let right = this.getRightChild(index);
        let left = this.getLeftChild(index)

        if (right !== -1 && this.array[right] > this.array[biggest]) {
            biggest = right
        }

        if (left !== -1 && this.array[left] > this.array[biggest]) {
            biggest = left
        }

        return biggest
    }

    getSmallestChild(index) {
        let smallest = index
        let right = this.getRightChild(index);
        let left = this.getLeftChild(index)

        if (right !== -1 && this.array[right] < this.array[smallest]) {
            smallest = right
        }

        if (left !== -1 && this.array[left] < this.array[smallest]) {
            smallest = left
        }

        return smallest
    }

    heapify(index) {
        while (index < this.array.length) {
            let child

            switch (this.type) {
                case 'min':
                    child = this.getSmallestChild(index)
                    break
                case 'max':
                    child = this.getBiggestChild(index)
                    break
            }

            if (child === index) return
            this.swap(index, child);
            index = child
        }
    }

    build() {
        let middle = Math.floor((this.array.length - 1) / 2)
        for (let i = middle; i >= 0; i--) {
            this.heapify(i)
        }
    }

    insert(element){
        this.array.push(element)
        let parent = this.getParent(this.array.length - 1)
        let child
        
        while (parent >= 0) {
            switch (this.type) {
                case 'min':
                    child = this.getSmallestChild(parent)
                    break
                case 'max':
                    child = this.getBiggestChild(parent)
                    break
            }
            if (child === parent) return
            this.swap(parent, child)
            parent = this.getParent(parent)
        }
    }

    delete() {
        if (this.array.length <= 1) {
            return this.array.pop()
        }

        let root = this.array[0]
        this.array[0] = this.array.pop()
        this.heapify(0)

        return root
    }

    sort() {
        let result = []
        let length = this.array.length

        for (let i = 0; i < length; i++) {
            result.push(this.delete())
        }
        
        this.array = result
        return this.array
    }
}


const heap = new Heap('min', [7, 5, 3, 8, 9, 1, 4, 2, 6])
console.log(heap.array)
heap.delete()
console.log(heap.array)
console.log(heap.sort())
heap.insert(0)
heap.insert(-1)
console.log(heap.array)

export default Heap

File: CPP/DS/Heap.cpp
#include <iostream>
#include "Array.cpp"

enum Type
{
    MIN = 0,
    MAX = 1
};

template <typename T>
class Heap
{
private:
    Array<T> array;
    Type type;

    int get_parent(int index)
    {
        return (index - 1) / 2;
    }

    int get_min(int parent)
    {
        int min = parent;
        int left = parent * 2 + 1;
        int right = parent * 2 + 2;

        if (left < array.size() && array[left] < array[min])
        {
            min = left;
        }

        if (right < array.size() && array[right] < array[min])
        {
            min = right;
        }

        return min;
    }

    int get_max(int parent)
    {
        int max = parent;
        int left = parent * 2 + 1;
        int right = parent * 2 + 2;

        if (left < array.size() && array[left] > array[max])
        {
            max = left;
        }

        if (right < array.size() && array[right] > array[max])
        {
            max = right;
        }

        return max;
    }

    void heapify(int index)
    {
        int parent = index;
        while (parent < array.size())
        {
            int child;
            switch (type)
            {
            case MIN:
                child = get_min(parent);
                break;
            case MAX:
                child = get_max(parent);
                break;
            }

            if (child == parent)
                return;

            array.swap(parent, child);
            parent = child;
        }
    }

    void build()
    {
        int middle = (array.size() - 1) / 2;
        for (int i = middle; i >= 0; i--)
        {
            heapify(i);
        }
    }

public:
    template <typename U>
    friend std::ostream &operator<<(std::ostream &os, const Heap<U> &heap);

    Heap(Type heap_type = MIN) : array(), type(heap_type)
    {
    }

    Heap(std::initializer_list<T> init_list, Type heap_type = MIN) : array(init_list), type(heap_type)
    {
        build();
    }

    void switch_type(Type heap_type)
    {
        if (heap_type != type)
        {
            type = heap_type;
            build();
        }
    }

    void insert(T element)
    {
        array.push(element);

        int child = array.size() - 1;
        while (child >= 0)
        {
            int parent = get_parent(child);
            bool swap;

            switch (type)
            {
            case MIN:
                swap = parent >= 0 && array[child] < array[parent];
                break;

            case MAX:
                swap = parent >= 0 && array[child] > array[parent];
                break;
            }

            if (!swap)
                return;

            array.swap(child, parent);
            child = parent;
        }
    }

    T *pop()
    {
        if (array.size() == 0)
        {
            return nullptr;
        }

        array.swap(0, array.size() - 1);
        T *element = array.pop();

        heapify(0);

        return element;
    }

    void print() const
    {
        array.print();
    }
};

template <typename T>
std::ostream &operator<<(std::ostream &os, const Heap<T> &heap)
{
    heap.print();
    return os;
}